### PR TITLE
Ensure shift instrinsic arguments match width of compiler-rt's.

### PR DESCRIPTION
**Pull request by cr1901**
This fixes [#520](https://github.com/rust-lang/compiler-builtins/issues/520). I tested this locally and confirmed this fixes my issue, along with the additional changes listed in the below sections _just for testing_.

I realized while working on this patch that as far as `compiler-builtins` is concerned, only the shift instrinsics are affected by 16-bit vs 32-bit mismatch. This dates back to a [patch](https://github.com/llvm/llvm-project/commit/4d41df64828195aa24cddc5d34d3f446ca7bb6d1) by @aykevl in 2020.

It is very possible that the msp430 backend generates code w/ incorrect width for other instrinsics changed in the above patch. However, I don't know a good way to sift these out w/ tests, Rust or otherwise. Perhaps https://github.com/rust-lang/rust/issues/107612 could be useful here? `mspdebug` has a simulator; it's [how](https://github.com/rust-lang/rust/issues/107612#issuecomment-1472377638) I was able to get a test case for #520.

## `compiler-builtins` patches

```diff
diff --git a/src/lib.rs b/src/lib.rs
index 71f249c..dabe076 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -14,6 +14,7 @@
 #![no_builtins]
 #![no_std]
 #![allow(unused_features)]
+#![allow(unexpected_cfgs)]
 // We use `u128` in a whole bunch of places which we currently agree with the
 // compiler on ABIs and such, so we should be "good enough" for now and changes
 // to the `u128` ABI will be reflected here.
```

Without this, for some reason Rust bails with `-D warnings`, due to https://github.com/rust-lang/rust/issues/96472 when I'm using `compiler-builtins` as a path dep (but not when using a crates.io dep... huh...).

## `rust` patches

```diff
diff --git a/Cargo.lock b/Cargo.lock
index 449f0c73588..c8a69ff7d5a 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -868,9 +868,7 @@ dependencies = [
 
 [[package]]
 name = "compiler_builtins"
-version = "0.1.87"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "f867ce54c09855ccd135ad4a50c777182a0c7af5ff20a8f537617bd648b10d50"
+version = "0.1.90"
 dependencies = [
  "cc",
  "rustc-std-workspace-core",
diff --git a/Cargo.toml b/Cargo.toml
index 15cbb2659c9..8fa9b040bfa 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -104,6 +104,12 @@ gimli.debug = 0
 miniz_oxide.debug = 0
 object.debug = 0
 
+[profile.release.package.core]
+opt-level = 's'
+
+[profile.release.package.alloc]
+opt-level = 's'
+
 [patch.crates-io]
 # See comments in `src/tools/rustc-workspace-hack/README.md` for what's going on
 # here
@@ -114,6 +120,7 @@ rustc-workspace-hack = { path = 'src/tools/rustc-workspace-hack' }
 rustc-std-workspace-core = { path = 'library/rustc-std-workspace-core' }
 rustc-std-workspace-alloc = { path = 'library/rustc-std-workspace-alloc' }
 rustc-std-workspace-std = { path = 'library/rustc-std-workspace-std' }
+compiler_builtins = { version="0.1.87", path = '../compiler-builtins' }
 
 [patch."https://github.com/rust-lang/rust-clippy"]
 clippy_lints = { path = "src/tools/clippy/clippy_lints" }
diff --git a/library/backtrace b/library/backtrace
--- a/library/backtrace
+++ b/library/backtrace
@@ -1 +1 @@
-Subproject commit 07872f28cd8a65c3c7428811548dc85f1f2fb05b
+Subproject commit 07872f28cd8a65c3c7428811548dc85f1f2fb05b-dirty
```

Use local `compiler-builtins` with my patch, work around some _unrelated_ LLVM assertion failures at `opt-level=3`.

### CODE REVIEW

**Comment by cr1901** [comment]
>https://github.com/rust-lang/compiler-builtins/actions/runs/4523089026/jobs/7965979554?pr=522#step:8:941

Interesting... well, `floatuntisf` [does use](https://github.com/rust-lang/compiler-builtins/blob/5511f3087255236680eefb862458ed2f90e11bb5/src/float/conv.rs#L52-L60) shifts, but without having access to the assembly, I can't tell why these conversions weren't a noop on 32-bit platforms and above. Ahhh well, not solving that tonight, will take a look when I have time.

**Comment by Amanieu** [comment]
Can you rebase now that #524 is merged, which should fix CI?

**Comment by cr1901** [comment]
@Amanieu Done, and CI is green :D!

File: src/int/shift.rs [hunk 1]
```diff
@@ -78,8 +78,8 @@ intrinsics! {
     #[avr_skip]
     #[maybe_use_optimized_c_shim]
     #[arm_aeabi_alias = __aeabi_llsl]
-    pub extern "C" fn __ashldi3(a: u64, b: u32) -> u64 {
-        a.ashl(b)
+    pub extern "C" fn __ashldi3(a: u64, b: core::ffi::c_uint) -> u64 {
+        a.ashl(b as u32)
```
Comment (aykevl) [inline src/int/shift.rs:82]: Drive by comment:

While this appears to follow the calling convention that LLVM uses, it does _not_ follow the calling convention that avr-gcc uses: https://godbolt.org/z/azq4749hd
`b` is only a single byte on avr-gcc.
(I also see now that the C implementation in compiler-rt has the same bug, but only realized it now).

Comment (cr1901) [reply src/int/shift.rs:82]: AIUI, these code paths are skipped when compiling for `avr`, precisely because there's a pass which gets rid of these types of shifts in LLVM. So I made no accommodations for its nonstandard calling conventions in this PR.

Comment (aykevl) [reply src/int/shift.rs:82]: > because there's a pass which gets rid of these types of shifts in LLVM

I know, because I wrote that pass ;)
But it only works for 32-bit values.`__ashldi3` is a 64-bit shift function, which is emitted as a regular compiler-rt call. Proof: https://godbolt.org/z/zqWfE6ndq

Anyway, it doesn't really matter as LLVM uses `i16` and not `i8` like avr-gcc does. This PR would only be buggy for avr-gcc, _right now_ it is safe with LLVM (but this could in theory change as `i8` is more efficient).

Comment (cr1901) [reply src/int/shift.rs:82]: Minddump for future-me follows:

>b is only a single byte on avr-gcc.

* According to [gcc docs](https://gcc.gnu.org/onlinedocs/gccint/Integer-library-routines.html), `__ashldi3`'s `b` argument is supposed to take an `int`, which must be at least 16 bits.  But indeed, your GCC code snippet is only sending a byte to  `__ashldi3`, while the LLVM version sends 16 bits (the `clr r17` line is clearing the top-most bits).
* On the other hand, `gcc`'s docs also says `__ashldi3` is supposed to return a `long`, and `__ashlti3` is the function that is supposed to return a `long long`. 
* On the other _other hand, `compiler-rt` doesn't seem to support the `t` versions of functions, and tends to hardcode a 64/32-bit int in places gcc's builtins do not<sup>1</sup>. In addition to `compiler-rt` uses `unsigned` in args where gcc's builtins are `signed`.

In other words, it seems like _both_ `gcc` `compiler-rt` builtins tend to do their own things and aren't directly compatible with each other. `avr` using a byte in for the `gcc` `__ashldi3` `b` arg, but 2 bytes for `compiler-rt`'s `__ashldi3` `b` arg is just one example of this. It's not going to be fun :P. Unifying `compiler-rt`'s and `gcc`'s builtins in terms of `compiler_builtins` might be necessary due to the `gcc` backend though.

1. The shift functions' `b` args notwithstanding :P.

Comment (cr1901) [reply src/int/shift.rs:82]: >But it only works for 32-bit values. __ashldi3 is a 64-bit shift function, which is emitted as a regular compiler-rt call.

How does this not break for Rust code, where all the shift functions are [skipped](https://github.com/rust-lang/compiler-builtins/blob/master/src/int/shift.rs#L72-L124) for AVR (`avr_skip`)?
